include/seqan3/io/stream/parse_condition.hpp

```diff
@@ -0,0 +1,1435 @@
+     *
+     * \details
+     *
+     * ### Complexity
+     *
+     * Constant
+     *
+     * ### Exception
+     *
+     * No-throw guarantee if the expression `std::is_same_v<char_t, char>`.
+     * Otherwise might throw, if the integer representation of the given wide character value is bigger than the
+     * biggest representable ASCII sign (> 127).
+     *
+     * ### Concurrency
+     *
+     * Thread-safe.
+     */
+    template <char_adaptation_concept char_t>
+    constexpr bool
+    operator()(char_t const & val) const noexcept(std::is_same_v<char_t, char>)
+    {
+        if constexpr (!std::is_same_v<char_t, char>)
+        {
+            if (static_cast<uint32_t>(val) > 127u)
+                throw parse_error{"The given value is out of the testable range."};
+        }
+
+        return (rng_beg <= val) && (val <= rng_end);
+    }
+};
+
+// ----------------------------------------------------------------------------
+// is_in_alphabet
+// ----------------------------------------------------------------------------
+
+/*!\brief Parse condition that checks if a given value is within the given alphabet `alphabet_t`.
+ * \ingroup stream
+ * \implements seqan3::detail::parse_condition_concept
+ * \tparam alphabet_t The alphabet type. Must satisfy the seqan3::alphabet_concept.
+ */
+template <alphabet_concept alphabet_t>
+struct is_in_alphabet : public detail::parse_condition<is_in_alphabet<alphabet_t>>
+{
+    //!\brief The message representing this condition.
+    static constexpr auto msg = constexpr_string{"is_in_alphabet<"} +
+                                constexpr_string{detail::get_display_name_v<alphabet_t>} +
+                                constexpr_string{">"};
+
+    /*!\brief Invokes the condition check for `val`.
+     * \tparam    char_t The type of the character to test. Must satisfy seqan3::char_adaptation_concept.
+     * \param[in] val    The value to test.
+     * \return `bool`; `true` if the condition is met, `false` otherwise.
+     *
+     * \details
+     *
+     * ### Complexity
+     *
+     * Constant
+     *
+     * ### Exception
+     *
+     * No-throw guarantee if the expression `std::is_same_v<char_t, char>`.
+     * Otherwise might throw, if the integer representation of the given wide character value is bigger than the
+     * biggest representable ASCII sign (> 127).
+     *
+     * ### Concurrency
+     *
+     * Thread-safe.
+     */
+    template <char_adaptation_concept char_t>
+    constexpr bool
+    operator()(char_t && val) const noexcept(std::is_same_v<char_t, char>)
+    {
+        if constexpr (!std::is_same_v<char_t, char>)
+        {
+            if (static_cast<uint32_t>(val) > 127u)
+                throw parse_error{"The given value is out of the testable range."};
+        }
+
+        // TODO(rrahn): implement to_upper / to_lower!
+        return to_char(assign_char(alphabet_t{}, val)) == std::toupper(static_cast<uint8_t>(val));
```

Comment:
This is also more tricky in reality. Right now our alphabets print upper case, but that isn't required anywhere. In fact, the `masked<dna5>` will have both lower and upper case as distinct characters and anyone else might make an alphabet that only accepts lower or only accepts upper case letters.

Let's discuss a better solution in person.

Reply:
That's true. I open an Issue for this.